src/SOS/Strike/strike.cpp

```diff
@@ -2464,7 +2466,27 @@ size_t FormatGeneratedException (DWORD_PTR dataPtr,
             WCHAR filename[MAX_LONGPATH] = W("");
             ULONG linenum = 0;
             if (bLineNumbers && 
-                SUCCEEDED(GetLineByOffset(TO_CDADDR(ste.ip), &linenum, filename, _countof(filename))))
+                // To get the source line number of the actual code that threw an exception, the IP needs 
+                // to be adjusted in certain cases. 
+                //
+                // The IP of the leaf node stack frame (i == 0) points to either:
+                //
+                // 1) The instruction that caused a hardware exception (div by zero, null ref, etc).
+                // 2) After the call instruction to an internal runtime function (FCALL like IL_Throw,
+                //    IL_Rethrow, JIT_OverFlow, etc.) that caused a software exception.
+                //
+                // It is case 2 that needs to adjust IP because these IL calls don't contain any other instructions
+                // after and point to next IL instruction and source line after the throw. We distinguish between #1
+                // and #2 by the bAsync flag which is set to true for hardware exceptions.
+                //
+                // The rest of the stack frame IPs also point to the instruction after the call. The difference 
+                // in this case is that these IL calls consist of more code after the call (at the minimum a nop
```

Comment:
A nop after a call isn't mandatory (at least not in release code). For example:

```
        static void Main(string[] args)
        {
            while (!Debugger.IsAttached) { }
            Foo();
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static void Foo()
        {
            Bar();
            Bar();
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static void Bar()
        {
        }
```

The disassembly of Foo() is:

```
            Bar();
00007FFB7E830BF0  push        rbp  
00007FFB7E830BF1  sub         rsp,20h  
00007FFB7E830BF5  lea         rbp,[rsp+20h]  
00007FFB7E830BFA  call        00007FFB7E82C698   // there is no nop after this call, if the IP points at the following instruction then it will resolve to the next line
            Bar();
00007FFB7E830BFF  call        00007FFB7E82C698  
00007FFB7E830C04  nop  
00007FFB7E830C05  lea         rsp,[rbp]  
00007FFB7E830C09  pop         rbp  
00007FFB7E830C0A  ret  
```

If the goal is to ensure that the line number of the source that made the call is always printed then it appears more robust to wind back the IP after all call instructions, not just the leaf frame.

Reply:
So basically we need to decrement IP for all stack frames not just leaf nodes.